File: python/oelMVCS/Core/Logger.py
# -*- coding: utf-8 -*-

from enum import Enum

class LogLevel(Enum):
    NONE = 0
    EXCEPTION = 1
    ERROR = 2
    WARNING = 3
    INFO = 4
    DEBUG = 5
    TRACE = 6
    ALL = 7

class Logger:
    def __init__(self):
        self.__level = LogLevel.ALL

    def Trace(self, _message:str, _args):
        self.__log(LogLevel.TRACE, self._trace, _message, _args)

    def Debug(self, _message:str, _args):
        self.__log(LogLevel.DEBUG, self._debug, _message, _args)

    def Info(self, _message:str, _args):
        self.__log(LogLevel.INFO, self._info, _message, _args)

    def Warning(self, _message:str, _args):
        self.__log(LogLevel.WARNING, self._warning, _message, _args)

    def Error(self, _message:str, _args):
        self.__log(LogLevel.ERROR, self._error, _message, _args)

    def Exception(self, _message:str, _args):
        if (self.__level < LogLevel.EXCEPTION):
            return
        self._exception(_ex);

    def _trace(self, _categoray:str, _message:str):
        pass

    def _debug(self, _categoray:str, _message:str):
        pass

    def _info(self, _categoray:str, _message:str):
        pass

    def _warning(self, _categoray:str, _message:str):
        pass

    def _error(self, _categoray:str, _message:str):
        pass

    def _exception(self, _categoray:str, _message:str):
        pass

    def __log(self, _level:LogLevel, _log, _message:str, _args):
        if (self.__level < _level):
            return

        message = str.format(_message, _args)
        categoray = str.format("oelMVCS::{}", getMethod())
        self._log(categoray, message)

    def getMethod(self) -> str:
        return ""

File: csharp/oelMVCS/Core/Logger.cs
/********************************************************************
     Copyright (c) XTechCloud
     All rights reserved.
*********************************************************************/

namespace XTC.oelMVCS
{
    public enum LogLevel
    {
        NONE,
        EXCEPTION,
        ERROR,
        WARNING,
        INFO,
        DEBUG,
        TRACE,
        ALL,
    }

    public class Logger
    {
        private System.Diagnostics.StackTrace stackTrace_ = new System.Diagnostics.StackTrace();

        public void setLevel(LogLevel _level)
        {
            level_ = _level;
        }

        public virtual void Trace(string _message, params object[] _args)
        {
            log(LogLevel.TRACE, this.trace, _message, _args);
        }

        public virtual void Debug(string _message, params object[] _args)
        {
            log(LogLevel.DEBUG, this.debug, _message, _args);
        }

        public virtual void Info(string _message, params object[] _args)
        {
            log(LogLevel.INFO, this.info, _message, _args);
        }

        public virtual void Warning(string _message, params object[] _args)
        {
            log(LogLevel.WARNING, this.warning, _message, _args);
        }

        public virtual void Error(string _message, params object[] _args)
        {
            log(LogLevel.ERROR, this.error, _message, _args);
        }

        public virtual void Exception(System.Exception _ex)
        {
            if (level_ < LogLevel.EXCEPTION)
                return;
            this.exception(_ex);
        }

        protected virtual void trace(string _categoray, string _message)
        {
        }

        protected virtual void debug(string _category, string _message)
        {
        }

        protected virtual void info(string _category, string _message)
        {
        }

        protected virtual void warning(string _category, string _message)
        {
        }

        protected virtual void error(string _category, string _message)
        {
        }

        protected virtual void exception(System.Exception _exception)
        {
        }

        private void log(LogLevel _level, System.Action<string, string> _log, string _message, params object[] _args)
        {
            if (level_ < _level)
                return;

            string message = "";
            if (null != _args && _args.Length > 0)
            {
                message = string.Format(_message, _args);
            }
            else
            {
                message = _message;
            }
            string? method;
            string? file;
            getMethod(out file, out method);
            string categoray = string.Format("{0}::{1}", file ?? "", method ?? "");
            _log(categoray, message);
        }

        private void getMethod(out string? _file, out string? _method)
        {
            try
            {
                System.Diagnostics.StackFrame sf = new System.Diagnostics.StackTrace().GetFrame(3);
                _method = sf.GetMethod().Name;
                _file = sf.GetMethod().DeclaringType.FullName;
            }
            catch (System.Exception)
            {
                _method = "";
                _file = "";
            }
        }

        private LogLevel level_ = LogLevel.ALL;
    }
}
